### PR TITLE
ENH: Add Moreau envelope

### PR DESCRIPTION
**Pull request by adler-j**
Initial work on Moreau envelope, it seems we need to add more properties. TODO:

- [x] Add more documentation
- [x] Add an example (e.g. showing that moreau envelope of l1 norm is huber norm)
- [x] Add test

### CODE REVIEW

**Comment by adler-j** [comment]
Ready for review.

File: odl/solvers/functional/default_functionals.py [hunk 1]
```diff
@@ -2101,6 +2102,87 @@ def __repr__(self):
...
+https://web.stanford.edu/~boyd/papers/pdf/prox_algs.pdf
+    """
+
+    def __init__(self, functional, sigma=1.0):
+        """Initialize an instance.
+
+        Parameters
+        ----------
+        functional : `Functional`
+            The functional ``f`` in the definition of the moreau envelope,
```
Comment (chongchenmath) [inline odl/solvers/functional/default_functionals.py]: Moreau

File: odl/solvers/functional/default_functionals.py [hunk 2]
```diff
@@ -2101,6 +2102,87 @@ def __repr__(self):
...
+    The Moraeu evelope is a way to smooth an arbitrary convex functional
+    such that its gradient can be computed given the proximal of the original
+    functional and has the same critical points as the functional.
+
+    It is also called the Moreau-Yosida regularization.
+
+    Note that the only computable property of the Moreau envelope is the
+    gradient, the functional itself cannot be evaluated efficiently.
+
+    See `Proximal Algorithms`_ for more information.
```
Comment (chongchenmath) [inline odl/solvers/functional/default_functionals.py]: redundant underscore

File: odl/solvers/functional/default_functionals.py [hunk 3]
```diff
@@ -2101,6 +2102,87 @@ def __repr__(self):
...
+    def __init__(self, functional, sigma=1.0):
+        """Initialize an instance.
+
+        Parameters
+        ----------
+        functional : `Functional`
+            The functional ``f`` in the definition of the moreau envelope,
+            the function to be smoothed.
+        sigma : positive float
+            The scalar ``sigma`` in the definition of the moreau envelope.
```
Comment (chongchenmath) [inline odl/solvers/functional/default_functionals.py]: same problem

File: odl/test/solvers/functional/default_functionals_test.py [hunk 4]
```diff
@@ -489,5 +489,36 @@ def test_separable_sum(space):
     assert func.convex_conj([x, y]) == l1.convex_conj(x) + l2.convex_conj(y)
 
 
+def test_moreau_envelope_l1():
+    """Test for the moreau envelope with L1 norm."""
```
Comment (chongchenmath) [inline odl/test/solvers/functional/default_functionals_test.py]: Moreau

File: odl/test/solvers/functional/default_functionals_test.py [hunk 5]
```diff
@@ -489,5 +489,36 @@ def test_separable_sum(space):
...
+                            [-1, 1, 1])
+
+    # Test with different sigma
+    smoothed_l1 = odl.solvers.MoreauEnvelope(l1, sigma=0.5)
+    assert all_almost_equal(smoothed_l1.gradient([0, 0.2, 0.7]),
+                            [0, 0.4, 1.0])
+
+
+def test_moreau_envelope_l2_sq(space, sigma):
+    """Test for the moreau envelope with l2 norm squared."""
```
Comment (chongchenmath) [inline odl/test/solvers/functional/default_functionals_test.py]: Moreau

File: odl/solvers/functional/default_functionals.py (hunk 2 above)
Comment (adler-j) [inline odl/solvers/functional/default_functionals.py]: Underscore is there to create a link

**Comment by adler-j** [comment]
Thanks for the review @chongchenmath! Will fix the comments.

For next time, you could use the new (since a few weeks) [pull request review](https://help.github.com/articles/about-pull-request-reviews/) feature, it's quite cool.

File: odl/solvers/functional/default_functionals.py [hunk 6]
```diff
@@ -2101,6 +2102,87 @@ def __repr__(self):
                                          self.__norm.outernorm.exponent,
                                          self.__norm.pwisenorm.exponent)
 
+
+class MoreauEnvelope(Functional):
+    """Moreau envelope of a convex functional.
+
+    The Moraeu evelope is a way to smooth an arbitrary convex functional
+    such that its gradient can be computed given the proximal of the original
```
Comment (chongchenmath) [inline odl/solvers/functional/default_functionals.py]: given ---> by

File: examples/solvers/lbfgs_tomography_tv.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+discr_phantom = odl.phantom.shepp_logan(reco_space, modified=True)
+
+# Create sinogram of forward projected phantom with noise
+data = ray_trafo(discr_phantom)
+data += odl.phantom.white_noise(ray_trafo.range) * np.mean(data) * 0.1
+
+# --- Set up optimization problem and solve --- #
+
+# Create data term ||Ax - b||_2^2 as composition of l2 norm squared
+# and the residual operator.
```
Comment (kohr-h) [inline examples/solvers/lbfgs_tomography_tv.py]: "squared L2 norm"
There's no residual operator, we removed it. Perhaps "composition of the squared L2 norm and the ray trafo translated by the data."

File: examples/solvers/lbfgs_tomography_tv.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+l2_norm = odl.solvers.L2NormSquared(ray_trafo.range)
+data_discrepancy = l2_norm * (ray_trafo - data)
+
+# Create regularizing functional || |grad(x)| ||_1 and smooth the functional
+# using the Moreau envelope.
+gradient = odl.Gradient(reco_space)
+l1_norm = odl.solvers.GroupL1Norm(gradient.range)
+
+# The parameter sigma controlls the strength of the regularization.
```
Comment (kohr-h) [inline examples/solvers/lbfgs_tomography_tv.py]: controls
Move this comment before the `gradient = ` line and put the code immediately after the block above.

File: examples/solvers/lbfgs_tomography_tv.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+smoothed_l1 = odl.solvers.MoreauEnvelope(l1_norm, sigma=0.03)
+
+# Compose with gradient in order to create the target functional
+regularizer = smoothed_l1 * gradient
```
Comment (kohr-h) [inline examples/solvers/lbfgs_tomography_tv.py]: Remove comment and put together with the rest above. There's not so much happening here that every line needs a comment.

File: examples/solvers/lbfgs_tomography_tv.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+
+# Create objective functional
```
Comment (kohr-h) [inline examples/solvers/lbfgs_tomography_tv.py]: total / overall / full objective functional

File: odl/solvers/functional/default_functionals.py [hunk 11]
```diff
@@ -2101,6 +2102,87 @@ def __repr__(self):
                                          self.__norm.outernorm.exponent,
                                          self.__norm.pwisenorm.exponent)
 
+
+class MoreauEnvelope(Functional):
```
Comment (kohr-h) [inline odl/solvers/functional/default_functionals.py]: add blank line

File: odl/solvers/functional/default_functionals.py [hunk 12]
```diff
@@ -2101,6 +2102,87 @@ def __repr__(self):
...
+    """Moreau envelope of a convex functional.
+
+    The Moraeu evelope is a way to smooth an arbitrary convex functional
```
Comment (kohr-h) [inline odl/solvers/functional/default_functionals.py]: typo "Moraeu" and "evelope"

File: odl/solvers/functional/default_functionals.py [hunk 13]
```diff
@@ -2101,6 +2102,87 @@ def __repr__(self):
...
+    such that its gradient can be computed given the proximal of the original
+    functional and has the same critical points as the functional.
```
Comment (kohr-h) [inline odl/solvers/functional/default_functionals.py]: Start new sentence: The new functional has the same critical points as the original.

File: odl/solvers/functional/default_functionals.py [hunk 14]
```diff
@@ -2101,6 +2102,87 @@ def __repr__(self):
...
+    .. math::
+        \mathrm{env}_{\\sigma  f}(x) =
+        \\inf_{y \\in \\mathcal{X}}
+        \\left\{ \\frac{1}{2 \\sigma} \| x - y \|_2^2 + f(y) \\right\}
+
+    The gradient of the envelope is given by
+
+    .. math::
+        [\\nabla \mathrm{env}_{\\sigma  f}](x) =
+        \\frac{1}{\\sigma} (x - \mathrm{prox}_{\\sigma  f}(x))
```
Comment (kohr-h) [inline odl/solvers/functional/default_functionals.py]: Nice!

File: odl/solvers/functional/default_functionals.py [hunk 15]
```diff
@@ -2101,6 +2102,87 @@ def __repr__(self):
...
+    """
+
+    def __init__(self, functional, sigma=1.0):
+        """Initialize an instance.
+
+        Parameters
+        ----------
+        functional : `Functional`
+            The functional ``f`` in the definition of the Moreau envelope,
+            the function to be smoothed.
```
Comment (kohr-h) [inline odl/solvers/functional/default_functionals.py]: "[...] envelope that is to be smoothed."

File: odl/solvers/functional/default_functionals.py [hunk 16]
```diff
@@ -2101,6 +2102,87 @@ def __repr__(self):
...
+        sigma : positive float
+            The scalar ``sigma`` in the definition of the Moreau envelope.
+            Higher sigma give a strong smoothing while smaller values give
```
Comment (kohr-h) [inline odl/solvers/functional/default_functionals.py]: Perhaps shorter: "Larger values mean stronger smoothing." The other direction is clear.

File: odl/solvers/functional/default_functionals.py [hunk 17]
```diff
@@ -2101,6 +2102,87 @@ def __repr__(self):
...
+                         linear=False)
+
+    @property
+    def functional(self):
+        """The functional that has been regularized."""
+        return self.__functional
+
+    @property
+    def sigma(self):
+        """Regularization constant, higher means stronger regularization."""
```
Comment (kohr-h) [inline odl/solvers/functional/default_functionals.py]: larger

File: examples/solvers/lbfgs_tomography_tv.py [hunk 18]
```diff
@@ -72,23 +72,20 @@
...
-# Create data term ||Ax - b||_2^2 as composition of l2 norm squared
-# and the residual operator.
+# Create data term ||Ax - b||_2^2 as composition of the squared L2 norm and the
+# ray trafo translated by the data.
 l2_norm = odl.solvers.L2NormSquared(ray_trafo.range)
 data_discrepancy = l2_norm * (ray_trafo - data)
 
 # Create regularizing functional || |grad(x)| ||_1 and smooth the functional
 # using the Moreau envelope.
+# The parameter sigma controlls the strength of the regularization.
```
Comment (kohr-h) [inline examples/solvers/lbfgs_tomography_tv.py]: controls

**Comment by adler-j** [comment]
Fixed typo, merge after CI.

**Pull request merged**